doc/contributor/style-guide.rst

```diff
@@ -0,0 +1,113 @@
+.. _flycheck-style-guide:
+
+=============
+ Style Guide
+=============
+
+This document describes our code style.  It tells you what to look for when
+making changes to Flycheck, or when reviewing pull requests.
+
+Features
+========
+
+Flycheck’s scope and focus is providing the infrastructure and foundations for
+on-the-fly syntax checking.  Flycheck provides the basics but deep integration
+with particular programming languages is best left to :ref:`separate packages
+<flycheck-extensions>`.
+
+Whether a feature is within the scope of Flycheck is the :ref:`maintainer’s
+<flycheck-maintainers>` judgement call.  Generally we reserve the right to
+reject any pull request for being out of scope.
+
+* Avoid a *disproportionate amount of code* for a single syntax checker or
+  language.  Look at the built-in checkers for judgement.  A syntax checker that
+  requires a lot more code than any built-in checker is likely to be rejected.
+
+* Avoid *deep integration* with a particular UI or completion framework.  Emacs’
+  standard is our standard: We will reject code that is tied to Helm or Counsel.
+
+* Likewise do not deviate from Emacs’ default behaviour too much.  Stick to
+  Emacs’ standard for key bindings, interactive functions, etc.
+
+Style
+=====
+
+.. important::
+
+   Generally, ``make check compile`` must pass on Emacs 25 or newer.  This
+   command checks for some style errors and byte-compile warnings and errors.
+
+* Indent with the default indentation rules.
+
+* Follow the :infonode:`(elisp)Programming Tips` for Emacs Lisp.
+
+* Whitespace:
+
+  * 80 characters per line.
+  * Avoid tabs and trailing spaces.
+
+* Naming:
+
+  * Prefix all variables and functions with the name of the containing library,
+    i.e. ``flycheck-`` for everything that is in :file:`flycheck.el`.
+
+  * End boolean predicates with ``-p``, i.e. ``flycheck-valid-checker-p``.
+
+* Avoid macros, and use them for syntax only.
+
+* Adhere to the :infonode:`(elisp)Key Binding Conventions`.  Particularly do not
+  define keys in Emacs’ reserved keymaps or in the :samp:`C-c {LETTER}` space
+  for user bindings.
+
+Libraries
+=========
+
+* Do **not** advise built-in or 3rd party functions and commands.
+
+* Do **not** redefine built-in or 3rd party functions, unless for compatibility,
+  but then copy the newer definition verbatim.
+
+* Do **not** use ``with-eval-after-load`` and similar functions.
```

Comment:
These are for user configuration only.  A package should not create an implicit dependency on other libraries via `with-eval-after-load`.  That's too much magic, and makes things hard to test.

If someone feels the need to do that we should either move that into an entirely separate package or add a hard `require`, depending on the circumstances.